### ISSUE TITLE
LASR with known camera intrinsics/extrinsics

### ISSUE THREAD

**Issue by ecmjohnson** [google/lasr#10]
Hello, I would like to run LASR with known camera intrinsics & extrinsics. I believe this is already implemented, but I'm having some trouble understanding how to accomplish this myself. The mechanisms seem to be two-fold: with the [`use_gtpose` option](https://github.com/google/lasr/blob/a22780b533079befd29d0bc5e0f9ca8b95f43873/optimize.py#L35) and [providing per-frame camera files](https://github.com/google/lasr/blob/a22780b533079befd29d0bc5e0f9ca8b95f43873/dataloader/vid.py#L57) (parsing code [here](https://github.com/google/lasr/blob/a22780b533079befd29d0bc5e0f9ca8b95f43873/dataloader/vidbase.py#L177)). Could you clarify the functionality of these mechanisms? I was unable to find an example that made use of either, but if I missed one or you have one, that would also be helpful.

Another thing that confuses me is the [scaling of the scale](https://github.com/google/lasr/blob/a22780b533079befd29d0bc5e0f9ca8b95f43873/nnutils/mesh_net.py#L241) (lol) when `use_gtpose` is set even though the focal length is assigned equivalently if the camera files are provided or not that makes me think these two mechanisms might have different purposes and I am incorrectly conflating them.

Any clarification you can provide would be much appreciated! Thanks!

**Comment by gengshan-y** [comment]
Hi Erik, the "with gt camera" option is not tested in the latest codebase. I believe some non-trivial modifications are needed. It would be the easiest to go from the synthetic spot data with the following suggestions.

First, you will need to pass `--use_gtpose` to the program call. 

Then, you will need to prepare camera files which is compatible with the [parsing code](https://github.com/google/lasr/blob/a22780b533079befd29d0bc5e0f9ca8b95f43873/dataloader/vidbase.py#L177). The files are stored as `"database/DAVIS/Camera/Full-Resolution/sequence-name/00xxx.txt"`, and contains values in the `cammat` array used in [synthetic data generation](https://github.com/google/lasr/blob/a22780b533079befd29d0bc5e0f9ca8b95f43873/scripts/render_syn.py#L192), which contains `[focal length, translation-x, translation-y, quaternion-w, quaternion-x, quaternion-y, quaternion-z, translation-z]`. Note the translation and rotation quaternions are coordinate transformations from object to camera space. We assumed the principle point to be (0,0) in the normalized device, and therefore did not record it.

What the [parsing code](https://github.com/google/lasr/blob/a22780b533079befd29d0bc5e0f9ca8b95f43873/dataloader/vidbase.py#L177) does is to transform the camera intrinsics from the raw image coordinate to the cropped and scaled image coordinates.

I believe the [scaling of the scale](https://github.com/google/lasr/blob/a22780b533079befd29d0bc5e0f9ca8b95f43873/nnutils/mesh_net.py#L241) by 10x is not correct and should be removed.

**Comment by gengshan-y** [comment]
I pushed a modified script for synthetic data generation, which should save camera files as well. Please let me know how it goes.

**Comment by ecmjohnson** [comment]
Thanks for adding those camera files. The files are generated and read without any issue. I think the 10x factor might be related to the focal length for the synthetic spot scene being 10 and the scaling due to cropping not modifying, but overwriting the file's focal length. (The focal lengths were set as `cam[0]=1./alp` and I think it should be `cam[0]/=alp` with the 10x factor in `mesh_net.py` removed.)

When training, there appear to be some issue. The replacement of the predictions with the GT in [this `if opts.use_gtpose` statement](https://github.com/google/lasr/blob/1d8404232d2e51c330bdf83416343eda0290bcc0/nnutils/mesh_net.py#L233) did not result in the same shapes, so I repeated the tensor to match the shapes which *seemed* to be the correct operation for the data to match. This was the only runtime error I encountered with `--use_gtpose`; however, it does not appear to be correct. The network never manages to learn to overfit the `scale` parameter (output always settles at `1e-12`) and the `depth` also appears to be incorrect. Additionally, I tried overriding these incorrect predictions in `predictor.py`; however, the results do not appear to be correct either way (see below), so I suspect there is some error in how I am overriding the camera parameters.

I am using the `ppoint` [computed in the code](https://github.com/google/lasr/blob/1d8404232d2e51c330bdf83416343eda0290bcc0/nnutils/mesh_net.py#L246), which is a bit under 10 for all frames, not zero as you mentioned above. Should I set the principle point to be zero for the synthetic spot dataset?

[The patch of my changes (I left the `print` statements to match logs below)](https://github.com/google/lasr/files/6983406/diff.txt)

[The output logs running spot3 with `use_gtpose` for the patch above](https://github.com/google/lasr/files/6983447/spot_use_gtpose.txt)

The extraction output running with the patch above using `--use_gtpose` option:

![spot3_use_gtpose](https://user-images.githubusercontent.com/6766142/129384297-f3958259-189f-452a-bca0-71e862146f0f.gif)

The extraction output when overriding the `scale` and `depth` predictions with GT values (10 and 10) in `predictor.py`:

![spot3_use_gtpose](https://user-images.githubusercontent.com/6766142/129382283-a2edcbf8-e3fe-4800-8c66-45e069d5e349.gif)

**Comment by gengshan-y** [comment]
Hi, could you try modifying the following arguments in the scripts/spot3.sh?
```
--use_gtpose
--n_bones 1 
--n_hypo 1
```
This works for me.

**Comment by gengshan-y** [comment]
I pushed a new commit, please checkout this [additional note](https://github.com/google/lasr#additional-notes).

**Comment by ecmjohnson** [comment]
Excellent! That does result in a much more convincing cow, although not quite as good as without GT camera parameters (due to use of single bone?).

![spot3-gtcam](https://user-images.githubusercontent.com/6766142/129539533-5cab7a4f-3fa8-47e4-89f3-a37d41d75ed6.gif)

Does that mean if I use my own data with GT camera parameters I will be constrained to a single bone? This would probably cause issues for articulating objects and I don't understand why the use of GT camera parameters constrains the number of bones and number of hypothesis symmetry planes. Also, would I only use the GT poses for the first optimization call like in 'spot3-gtcam.sh`?

**Comment by gengshan-y** [comment]
The problem is likely due to the small number of views (this example only uses 3 views) and lack of smoothness constraint . The shape should be better if you increase it to ~20 views or use a larger weight for [laplacian smoothness ](https://github.com/google/lasr/blob/25722b724e02ee1ddc0dce5ffd14b4887b8f12d2/nnutils/mesh_net.py#L451).

It should be easy to modify the `use_gtpose` option to be compatible with more than one bone. My guess is that your previous modification repeated the camera predictions in the wrong way. For instance, you only want to modify the root body pose to be the ground-truth poses, but not the bone transformations, which ground-truth is not available. You would only want to repeat in the hypothesis axis to make the shape match.

Again, the `use_gtpose` option is not supported by this repo and I don't have cycles to push a commit that fully address the problem. But feel free to follow up if you want to continue on it.

**Comment by ecmjohnson** [comment]
Sorry to return to this after such an extensive absence, but I was wondering if you could clarify for me why we only pass `--use_gtpose` to the first optimization call for the spot sequence? How would this scale to the more complex datasets using 5 optimization steps (e.g. `--use_gtpose` for first 4 optimizations and then `--nouse_gtpose` for the last)?

Also, should the produced per-frame `cam*.txt` files match those provided (if the format is converted of course)? For the spot dataset this doesn't appear to be the case.

**Comment by gengshan-y** [comment]
The pose CNN is trained with the provided camera poses in the first optimization, and that's why it still works without passing `--use_gtpose` to the 2nd optimization. If it's desirable to always use GT cameras (e.g., the cameras are very accurate), you could pass `--use_gtpose` to all optimizations (I haven't tested).

Can you elaborate on the second point? The result of which optimization does not match? Is it off a lot?

**Comment by ecmjohnson** [comment]
I was just wondering why the output `cam*.txt` files (produced by the `extract.sh` script) always include no rotation or translation. I see [why that is the case](https://github.com/google/lasr/blob/25722b724e02ee1ddc0dce5ffd14b4887b8f12d2/nnutils/predictor.py#L277), but I don't understand it. Is the root rotation and translation removed from the meshes when they are extracted (ie. they are aligned with the camera)?

**Comment by ecmjohnson** [comment]
An additional question about those output `cam*.txt` files: Why are the written out focal length and [principal point scaled by 128](https://github.com/google/lasr/blob/25722b724e02ee1ddc0dce5ffd14b4887b8f12d2/nnutils/predictor.py#L188)? Should I not be looking at these files as the final optimized cameras?

**Comment by gengshan-y** [comment]
>  Is the root rotation and translation removed from the meshes when they are extracted (ie. they are aligned with the camera)?

Your understanding is correct. 

The values written to cam*.txt should correspond to the focal length and principal points in pixel units (before cropping to object bounding box). If I remember correctly, the focal length predicted by the network correspond to the true value (before cropping) but scaled by 1 /128.
